### PR TITLE
Enable setting the module name of a class

### LINKED ISSUE #474: Generated classes are part of the builtin module
Related to the pickle support #100 , as mentioned by @althonos ,
> PyO3 exposes all classes as part of the builtins module, because the import mechanism has not been properly implemented, so pickle tries to use builtins.MyClass and fails with the error you reported.

This is also discussed in the following test, 

https://github.com/PyO3/pyo3/blob/9f2afbc3cc6cceca8d9bc2fdfd3236032e4b9577/tests/test_class_basics.rs#L70-L73

Once `__module__` it is set to `builtin`, it cannot be changed in an obvious way (at least I have not found one). Using [the example class](https://github.com/PyO3/pyo3/blob/master/guide/src/class.md#constructor),
```py
       cls = MyClass()
       cls.__module__ = "correct_name"
E   TypeError: can't set attributes of built-in/extension type 'MyClass'
```

it would be nice if there way some way to manually set the module name.

Also setuptools-rust does have the module name provided by the user for each `RustExtension` , but doesn't seem to be assigning it?

### PR DESCRIPTION
**Pull request by Alexander-N**
Didn't implement setting it automatically to avoid unexpected behavior. Pickling is possible but very awkward.

#100
Closes #474

### CODE REVIEW

File: tests/test_various.rs [hunk 1]
```diff
@@ -117,3 +118,58 @@ fn pytuple_pyclass_iter() {
...
+}
+
+fn add_module(py: Python, module: &PyModule) {
+    py.import("sys")
+        .expect("failed to import python sys module")
+        .dict()
+        .get_item("modules")
+        .expect("failed to get python modules dictionary")
+        .downcast_mut::<PyDict>()
+        .expect("failed to turn sys.modules into a PyDict")
```
Comment (kngwyu) [inline tests/test_various.rs]: Here we call `sys.modules` and ... can it fail?
I prefer to `unwrap` if the operation doesn't fail.

**Comment by kngwyu** [comment]
Thank you for investigating the problem!
Looks it's the correct solution to me.

File: src/type_object.rs [hunk 2]
```diff
@@ -315,6 +309,13 @@ where
 
     type_object.tp_base = base_type_object;
 
+    let name = match module_name {
+        Some(module_name) => format!("{}.{}", module_name, T::NAME),
+        None => T::NAME.to_string(),
+    };
+    let name = CString::new(name).expect("Module name/type name must not contain NUL byte");
```
Comment (lycantropos) [inline src/type_object.rs]: *NULL ?

Comment (althonos) [reply src/type_object.rs]: [NUL is valid in that context](https://en.wikipedia.org/wiki/Null_character) :wink:

**Pull request merged**